subprojects/normalization-java/src/main/java/org/gradle/api/internal/changedetection/state/AbstractLineEndingAwareHasher.java

```diff
@@ -0,0 +1,200 @@
+/*
+ * Copyright 2021 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.changedetection.state;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.primitives.Bytes;
+import org.gradle.internal.fingerprint.hashing.ZipEntryContext;
+import org.gradle.internal.hash.HashCode;
+import org.gradle.internal.hash.Hashing;
+import org.gradle.internal.hash.PrimitiveHasher;
+
+import javax.annotation.Nullable;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Iterator;
+import java.util.Optional;
+import java.util.function.Supplier;
+
+public abstract class AbstractLineEndingAwareHasher {
+    protected Optional<HashCode> hashContent(ZipEntryContext zipEntryContext) throws IOException {
+        return zipEntryContext.getEntry().isDirectory() ? Optional.empty() : zipEntryContext.getEntry().withInputStream(this::hashContent);
+    }
+
+    private Optional<HashCode> hashContent(InputStream inputStream) throws IOException {
+        return new LineEndingAwareResourceHasher.LineEndingAwareInputStreamHasher().hash(inputStream);
+    }
+
+    protected Optional<HashCode> hashContent(File file) throws IOException {
+        try (FileInputStream inputStream = new FileInputStream(file)) {
+            return hashContent(inputStream);
+        }
+    }
+
+    /**
+     * Provides convenience methods that allow providing a supplier that throws an exception.
+     */
+    protected static class Suppliers {
```

Comment:
This is all so we can add the call to `delegate.hash()` in an `orElseGet()` call in the hashers.  Because it throws `IOException` we can't just pass it in a `Supplier`.  The other option would be to do something like `isPresent() ? get() : delegate.hash()` in those methods.  I think this is only worthwhile if we adapt this into some sort of utility class.  However, this project doesn't currently depend on any other project where it might make sense to put such a thing.  @lptr Do you think this is worth keeping around?